### DISCUSSION TITLE
How to configure eglot to detect automatically the current running Python virtual environment?

### DISCUSSION THREAD

**Discussion by parvizfarnia** [joaotavora/eglot#1334] [Q&A]
Hello

I use `Emacs 29.1` and `Python 3.12.0`. So far, I have been using `elpy` in Emacs for Python development but given the project is no more maintained by its main developer, I switched to `eglot/LSP` and currently I am learning how to configure my environment. By watching a few YouTube tutorials and also instructions available in the documentation of [Python LSP Server](https://github.com/python-lsp/python-lsp-server), I managed to make work most of the things that I need on a daily basis.

However, I think that `eglot` has its own Python interpreter (or maybe even a virtual environment as I still don't know in details how it functions). For example, I have a running virtual environment in my terminal based on Python 3.12.0. Within the very same terminal running this virtual environment, I run Emacs in background. But if I start a Python process inside Emacs, it runs on Python 3.9. Therefore, it seems to me that eglot ignores my current running virtual environment.

I would like to know whethere there is any configuration to add to my Emacs init file to force eglot to take into account the current running Python virtual environment as the only and single environment (and also interpreter) for Python.

Thanks in advance.

**Comment by joaotavora** [comment]
>Within the very same terminal running this virtual environment, I run Emacs in background. 

This means it's probably not seeing the `PATH`, `PYTHONPATH` ,etc that your venv declares.  It's only if you launch Emacs as a child process to the shell that is running _within_ that Terminal, that you would make it inherit those venv settings.  

But you probably don't want to do that.

> However, I think that eglot has its own Python interprete

Eglot absolutely does not have an Python interpreter.  In fact Eglot know next to 0 about Python or any language. But if you want Eglot (and generally Emacs)  to work correctly for Python projects that depend on a virtual environment, my recommendation is to use something like https://github.com/porterjamesj/virtualenvwrapper.el.  After installing it, I use `M-x venv-workon` to select the virtual environment and `M-x eglot` , `M-x python` and even things like `M-x gud` work correctly.

**Reply by yagossc** [reply]
I'm sorry for necro-bumping this question, but if you install pylsp via pip in a virtual environment, run it and pass the `host:port` combination to `eglot` it works fine.
So, from within a venv:
```
pip install "python-lsp-server[all]"
pylsp --host 127.0.0.1 --port 9993 -v --tcp
```

This works for me.

> Note: if you jump to a definition within another "project", you'll have to run eglot and pass pylsp informations to it again, so this is not a "best practive".
